admin/create-theme/theme-utils.php

```diff
@@ -185,4 +185,55 @@ public static function is_valid_screenshot( $file ) {
 		return 0;
 	}
 
+	public static function is_valid_screenshot_file( $file_path ) {
+		return Theme_Utils::get_screenshot_file_extension( $file_path ) !== null;
+	}
+
+	public static function get_screenshot_file_extension( $file_path ) {
+		$allowed_screenshot_types = array(
+			'png'  => 'image/png',
+			'gif'  => 'image/gif',
+			'jpg'  => 'image/jpeg',
+			'jpeg' => 'image/jpeg',
+			'webp' => 'image/webp',
+			'avif' => 'image/avif',
+		);
+		$filetype                 = wp_check_filetype( $file_path, $allowed_screenshot_types );
+		if ( in_array( $filetype['type'], $allowed_screenshot_types, true ) ) {
+			return $filetype['ext'];
+		}
+		return null;
+	}
+
+	public static function process_screenshot( $file_path ) {
+		$new_screeenshot_id      = attachment_url_to_postid( $file_path );
+		$new_screenshot_metadata = wp_get_attachment_metadata( $new_screeenshot_id );
+		$upload_dir              = wp_get_upload_dir();
+
+		$new_screenshot_location = path_join( $upload_dir['basedir'], $new_screenshot_metadata['file'] );
+
+		// TODO: We're not actually processing anything.  The image might be of the wrong dimensions, too large, etc.
+		// We're just getting the file location so that we can copy it.
+		// Should we process it?  Or just enforce that the screenshot is the correct size?
+
+		return $new_screenshot_location;
+	}
+
+	public static function replace_screenshot( $new_screenshot_path ) {
+		if ( ! Theme_Utils::is_valid_screenshot_file( $new_screenshot_path ) ) {
+			return new \WP_Error( 'invalid_screenshot', __( 'Invalid screenshot file', 'create-block-theme' ) );
+		}
+
+		// Clean up, resize or compress the screenshot if necessary
+		$processed_screenshot_location = Theme_Utils::process_screenshot( $new_screenshot_path );
+
+		// Remove the old screenshot
+		unlink( path_join( get_stylesheet_directory(), wp_get_theme()->get_screenshot( 'relative' ) ) );
+
+		// Copy the new screenshot
+		$new_screenshot_filetype = Theme_Utils::get_screenshot_file_extension( $new_screenshot_path );
+		$new_location            = path_join( get_stylesheet_directory(), 'screenshot.' . $new_screenshot_filetype );
+		copy( $processed_screenshot_location, $new_location );
```

Comment:
[The theme handbook](https://developer.wordpress.org/themes/releasing-your-theme/required-theme-files/) states:

> In the WordPress.org theme directory, the screenshot acts as a visual indicator of what your theme looks like. It is visible both in the web view and in the admin dashboard. **The screenshot must not be bigger than 1200 x 900px**. 

Therefore, how about resizing the file to 1200 x 900px or less? It may be better to resize _after_ copying so as not to affect the original media.

I haven't tested it, but an approach like the one below might work?


```php
copy( $processed_screenshot_location, $new_location );

$image_editor = wp_get_image_editor( $new_location );
// Resize and crop image
$image_editor->resize( 1200, 900, true );
// Overwrite new screenshot
$image_editor->save();
```

Reply:
I did something very similar; thanks for getting me started. 
The editor copied the image and I pointed it at a new location on the save action so it was all one action.